src/Graph.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+namespace Spatie\SchemaOrg;
+
+use ReflectionClass;
+use BadMethodCallException;
+use InvalidArgumentException;
+
+/**
+ * @mixin Schema
+ */
+class Graph extends BaseType
+{
+    /** @var array */
+    protected $hidden = [];
+
+    public function __call(string $method, array $arguments)
+    {
+        if (is_callable([Schema::class, $method])) {
+            $type = (new ReflectionClass(Schema::class))->getMethod($method)->getReturnType();
+
+            $schema = $this->getOrCreate($type);
+
+            if (isset($arguments[0]) && is_callable($arguments[0])) {
+                call_user_func($arguments[0], $schema, $this);
+
+                return $this;
+            }
+
+            return $schema;
+        }
+
+        throw new BadMethodCallException(sprintf('The method "%" does not exist on class "%s".', $method, get_class($this)));
+    }
+
+    public function add(Type $schema)
+    {
+        $type = get_class($schema);
+        if ($this->has($type)) {
+            throw new InvalidArgumentException(sprintf('The graph already has an item of type "%s".', $type));
+        }
+
+        return $this->set($schema);
+    }
+
+    public function has(string $type): bool
+    {
+        return $this->offsetExists($type);
+    }
+
+    public function set(Type $schema)
+    {
+        return $this->setProperty(get_class($schema), $schema);
+    }
+
+    public function get(string $type): Type
+    {
+        if (! $this->has($type)) {
+            throw new InvalidArgumentException(sprintf('The graph does not have an item of type "%s".', $type));
+        }
+
+        return $this->getProperty($type);
+    }
+
+    public function getOrCreate(string $type): Type
+    {
+        if (!is_subclass_of($type, Type::class)) {
+            throw new InvalidArgumentException(sprintf('The given type "%s" is not an instance of "%s".', $type, Type::class));
+        }
+
+        if (!$this->has($type)) {
+            $this->set(new $type());
+        }
+
+        return $this->get($type);
+    }
+
+    public function hide(string $type)
```

Comment:
I think It should add the return type hint named `self`.

Reply:
I don't have good experience with `self` return types. If it comes to overriding you can't use `self` but have to use the extended classname. 😣
I'm open for the CR but want to share my thoughts why it hasn't happened. Even if I think that this class won't be extended/overriden this much.

Reply:
@Gummibeer, my confused problem is about why some methods misses the `return type hint`.

And I just think they should be added :).

Reply:
https://github.com/spatie/schema-org/pull/62/commits/452858e3b39ac06fba4ae1e5dfc6c94d87aa4082